index.rst

```diff
@@ -1,59 +1,367 @@
+
+At the same time, we will use filesystem access controls to protect shared and per-user files, and we plan to implement some checks in the Butler client itself to make it at least extremely difficult to *accidentally* cause problems.
+
+.. warning::
+
+   NEVER use ``psql`` or other direct-SQL clients (e.g. the Python DBAPI or SQLAlchemy) to perform write operations in the repository database.
+   These can corrupt the data repository, and we have essentially no way to guard against them.
+
+   It should not be necessary in the long term to ever use direct SQL access even for read access; the SQL schema is *not* considered a public interface - but we recognize that this may be necessary for debugging for a while.
+   This can be ensured by running::
+
+      SET SESSION CHARACTERISTICS AS TRANSACTION READ ONLY;
+
+   at the start of the session.
+
+   If you have to do this (and not at the prompting of a middleware team member trying to help diagnose a problem), please also create a ticket explaining what you wanted to do that couldn't be done with butler tools, so we can address that feature gap.
+
+This proposal specifies filesystem access controls in terms of a number of high-level "roles" that certain operators or developers may temporarily opt in to via ``su`` or special setuid tools.
+Usually these roles will be used only to create subdirectories that are owned directly by the true user.
+How to map these to users, groups, and filesystem, directory, or file-level permissions in detail is something I'd prefer to leave to the system administrators.
+All directories in ``/datasets`` will be world-readable.
+
+Within each ``repo_<YYYYMMDD>`` repository directory:
+
+ - Regular users will always have write access to their own ``u/<user>`` directory.
+
+ - Production operators will have access to an ``execution`` role that can write to ``runs`` and ``<instrument>/runs`` (usually just used to create an owned subdirectory).
+
+ - Production operators and certain CPP team members will have access to a ``calibs`` role that can write to all ``<instrument>/calib`` directories (usually just used to create an owned subdirectory).
+
+ - Production operators and science pipelines developers who regularly ingest raw data for one or more instruments will have access to per-instrument ``<instrument>_raw`` roles that have write access to the ``<instrument>/raw`` and any auxiliary-data-collection subdirectories (e.g. ``HSC/masks``).  At present, these roles would have to be used directly whenever ingesting new raws, not just to create subdirectories for them, but we may be able to improve this in the future.  For external instruments, this role would also ideally provide a way to get write access (at least for writing new files) to the ``/external-raw`` filesystem (though not necessarily at that mount point).
```

Comment:
(split paragraph by sentence?)

"At present, these roles would have to be used directly whenever ingesting new raws" - why is this a bad thing?

Reply:
> (split paragraph by sentence?)

:+1:

> "At present, these roles would have to be used directly whenever ingesting new raws" - why is this a bad thing?

If the implementation of these access controls involves sudo-ing to some special account, I'd like to minimal how much has to be done as that account.  That's both to minimize the damage caused by accidents, and to let people use as much of their own custom environments and convenience tooling when performing nontrivial tasks.